openwisp_notifications/notification_types.py

```diff
@@ -0,0 +1,78 @@
+from django.core.exceptions import ImproperlyConfigured
+from openwisp_notifications.settings import ADDITIONAL_NOTIFICATION_TYPES
+
+DEFAULT_NOTIFICATION_TYPES = {
+    'default': {
+        'level': 'info',
+        'verb': 'default verb',
+        'name': 'Default Type',
+        'email_subject': '[openwisp-notificaton] Default Notification Subject',
+    },
+}
+
+
+def deep_merge_dicts(dict1, dict2):
+    result = dict1.copy()
+    for key, value in dict2.items():
+        if isinstance(value, dict):
+            node = result.get(key, {})
+            result[key] = deep_merge_dicts(node, value)
+        else:
+            result[key] = value
+    return result
+
+
+def get_notification_types():
+    notification_types = deep_merge_dicts(
+        DEFAULT_NOTIFICATION_TYPES, ADDITIONAL_NOTIFICATION_TYPES
+    )
+    validate_notification_type(notification_types)
+    return notification_types
+
+
+def get_notification_choices():
+    notification_types = NOTIFICATION_TYPES
+    choices = []
+    for key in sorted(notification_types.keys()):
+        name = notification_types[key].get('name', key)
+        choices.append((key, name))
+    return choices
+
+
+def validate_notification_type(notification_type):
+    for key, options in notification_type.items():
+        assert 'level' in options
+        assert 'verb' in options
+
+
+def register_notification_type(notification_type):
```

Comment:
please document the usage of this function.

Test it in openwisp-monitoring to register a notification type there and ensure the notification type appears in the available choices.

**Addendum**

I've done a quick test:

```python
# ... omitting code ...
def ready(self):
    from openwisp_notifications.notification_types import register_notification_type
    register_notification_type({
        'monitoring.device.up': {
            'level': 'info',
            'verb': 'up',
            'name': 'UP',
            'email_subject': '[UP] test',
        }
    })
    register_notification_type({
        'monitoring.device.down': {
            'level': 'warning',
            'verb': 'down',
            'name': 'DOWN',
            'email_subject': '[DOWN] test',
        }
    })
```

Seems to work:

```
$ ./manage.py shell_plus
# Shell Plus Model Imports
# ... omitting output ...
from openwisp_notifications.models import Notification, NotificationUser
# ... omitting output ...

In [1]: Notification.type.field.choices                                                                 
Out[1]: 
[('default', 'Default Type'),
 ('monitoring.device.up', 'UP'),
 ('monitoring.device.down', 'DOWN')]
```

Cool! :+1: :clap: 

What I would change is the syntax:

```python
register_notification_type('monitoring.device.up', {
    'level': 'info',
    'verb': 'up',
    'name': 'UP',
    'email_subject': '[UP] yo',
})
```

Reply:
One more thing, see how django implements `register` and `unregister` for the admin:

https://github.com/django/django/blob/master/django/contrib/admin/sites.py#L88-L147

Reply:
> What I would change is the syntax:
> 
> ```python
> register_notification_type('monitoring.device.up', {
>     'level': 'info',
>     'verb': 'up',
>     'name': 'UP',
>     'email_subject': '[UP] yo',
> })
> ```

I think an iterable will make this syntax a bit messy. The current syntax of having `dict` of `dict` seems more appropriate. It also resembles with definition of `OPENWISP_NOTIFICATION_TYPES`. No doubt the suggested one looks very cleaner.